### PR TITLE
Feature/graceful shutdown

### PR DESCRIPTION
**Pull request by alpaka-alpachino**

### CODE REVIEW

**Comment by radikh** [comment]
Conflicts are still here.

**Comment by alpaka-alpachino** [comment]
> Conflicts are still here.

They are new, but will fix in few minutes.

**Comment by alpaka-alpachino** [comment]
> Conflicts are still here.

Resolved :'D

File: cmd/shutdown.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		// Handle extra services here (like closing database, etc.)
+
+		if err := srv.Shutdown(ctx); err != nil {
+			log.Fatalf("Server shutdown failed:%+v", err)
+		}
+
+		cancel()
+	}()
+
+	done <- true
```
Comment (radikh) [inline cmd/shutdown.go]: For such kind of channels it's even better to close them instead of sending values.

File: cmd/main.go [hunk 2]
```diff
@@ -2,12 +2,26 @@ package main
...
 )
 
 func main() {
-	for {
-		log.Println(" [INFO] App is running.")
-		time.Sleep(5 * time.Second)
-	}
+
+	sigs := make(chan os.Signal, 1)
+	done := make(chan bool, 1)
```
Comment (radikh) [inline cmd/main.go]: Do you really need buffered channels?

File: cmd/shutdown.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	fmt.Println("Recieved", sig, "signal")
+
+	timeout := 10 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+
+	//Replace srv value with your server
+	srv := &http.Server{
+		Addr: ":8080",
+	}
+
```
Comment (radikh) [inline cmd/shutdown.go]: We don't need to stop only http service but also close database, consul, kafka connections, and finish all transmissions and processing that is going on in application. Try to make this code also capable to handle other components.

File: cmd/shutdown.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	"os"
+	"time"
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func GracefulShutdown(sigs <-chan os.Signal, done chan<- bool) error {
+	sig := <-sigs
+	fmt.Println("Recieved", sig, "signal")
+
+	timeout := 10 * time.Second
```
Comment (radikh) [inline cmd/shutdown.go]: This value will better to be configured.

File: cmd/main.go [hunk 5]
```diff
@@ -2,12 +2,26 @@ package main
...
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		err := GracefulShutdown(sigs, done)
+		if err != nil {
+			log.Fatalf("Service graceful shutdown failed: %v", err)
+		}
+	}()
```
Comment (radikh) [inline cmd/main.go]: You don't need a goroutine here, just wait until GracefulShutdown finishes.

File: cmd/main.go (hunk 2 above)
Comment (alpaka-alpachino) [reply cmd/main.go]: Oh, you are right, there I don't need them

File: cmd/shutdown.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		log.Fatalf("Server shutdown failed:%+v", err)
+	}
+	kafka := kafka.NewConn(nc, "topic", 0)
+
+	client, _ := api.NewClient(api.DefaultConfig())
+	consul, err := connect.NewService("my-service", client)
+	if err != nil {
+		log.Fatalf("Server shutdown failed:%+v", err)
+	}
```
Comment (crnopster) [inline cmd/shutdown.go]: Do we need all of this?

File: cmd/shutdown.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+	otherComponents := structForClose{
+		components: []closer{
+			//Put connection variables here
+			db,
+			kafka,
+			consul,
+		},
+	}
```
Comment (crnopster) [inline cmd/shutdown.go]: We must use this in main

File: cmd/shutdown.go (hunk 3 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Made useful structure for that)

File: cmd/shutdown.go (hunk 4 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Got it)

File: cmd/shutdown.go (hunk 6 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Moved it to tests)

File: cmd/shutdown.go (hunk 7 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Yep)

File: cmd/shutdown.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	"context"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+type closer interface {
```
Comment (dmitriyminer) [inline cmd/shutdown.go:13]: Unnecessary type declaration, you can use `io.Closer`

File: cmd/shutdown.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	Close() error
+}
+
+type structForClose struct {
```
Comment (dmitriyminer) [inline cmd/shutdown.go:17]: You can use slice of io.Closer directly

File: cmd/shutdown.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	log.Println("Recieved", sig, "signal")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+
+	select {
+
+	case <-ctx.Done():
+		log.Println(ctx.Err())
+
+	case <-time.After(10 * time.Second):
```
Comment (dmitriyminer) [inline cmd/shutdown.go:38]: Magic number, it should be as a constant, for example `shutdownTimeout`

File: cmd/shutdown_test.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	"testing"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/connect"
+)
+
+func Test_structForClose_GracefulShutdown(t *testing.T) {
+	signal := make(chan bool)
+
+	client, _ := api.NewClient(api.DefaultConfig())
```
Comment (dmitriyminer) [inline cmd/shutdown_test.go:15]: You should check an error and call t.Fatal if error exists

File: cmd/shutdown_test.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+		log.Fatalf("Server shutdown failed:%+v", err)
+	}
+
+	db := structForClose{
+		components: []closer{
+			postgresDB,
+		},
+	}
+
+	type args struct {
```
Comment (dmitriyminer) [inline cmd/shutdown_test.go:38]: Unnecessary type declaration

File: cmd/shutdown.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+type closer interface {
+	Close() error
+}
+
+type structForClose struct {
+	components []closer
+}
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func (closing *structForClose) GracefulShutdown(done chan<- bool) error {
```
Comment (dmitriyminer) [inline cmd/shutdown.go:22]: It should be unexported

Comment (dmitriyminer) [reply cmd/shutdown.go:22]: WDYT about changing implementation from method to function without new type declaration?

File: cmd/shutdown.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+	sigs := make(chan os.Signal)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	sig := <-sigs
+	log.Println("Recieved", sig, "signal")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (dmitriyminer) [inline cmd/shutdown.go:31]: Magic number

File: cmd/shutdown.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+	default:
+		cancel()
+	}
+
+	// Please, replace srv with your server.
+	server := &http.Server{
+		Addr: ":8080",
+	}
+
+	defer func() {
```
Comment (dmitriyminer) [inline cmd/shutdown.go:51]: It shouldn't be in defer

**Pull request closed without merging**